llvm_passes/HipPromoteInts.cpp

```diff
@@ -0,0 +1,309 @@
+#include "HipPromoteInts.h"
+#include "llvm/IR/IRBuilder.h"
+#include "llvm/IR/Instructions.h"
+#include "llvm/Support/Debug.h"
+#include "llvm/Support/raw_ostream.h"
+
+#define DEBUG_TYPE "hip-promote-ints"
+
+/**
+ * This pass promotes integer types to the next standard bit width.
+ * During optimization of loops, LLVM generates non-standard integer types
+ * such as i33 or i56
+ * 
+ * __global__ void testWarpCalc(int* debug) {
+    int tid = threadIdx.x;
+    int bid = blockIdx.x;
+    int globalIdx = bid * blockDim.x + tid;
+    
+    // Optimizations on this loop will generate i33 types.
+    int result = 0;
+    for(int i = 0; i < tid + 1; i++) {
+        result += i * globalIdx;
+    }
+    
+    // Store using atomic operation
+    atomicExch(&debug[globalIdx], result);
+}
+ * 
+ * https://github.com/KhronosGroup/SPIRV-LLVM-Translator/issues/2823
+ */
+
+using namespace llvm;
+
+bool HipPromoteIntsPass::isStandardBitWidth(unsigned BitWidth) {
+  // TODO: 128 is not a standard bit width, will handle later as it's more complex than simply promoting
+  return BitWidth == 1 || BitWidth == 8 || BitWidth == 16 || BitWidth == 32 || BitWidth == 64 || BitWidth == 128;
+}
+
+unsigned HipPromoteIntsPass::getPromotedBitWidth(unsigned Original) {
+  if (Original <= 8) return 8;
+  if (Original <= 16) return 16;
+  if (Original <= 32) return 32;
+  return 64;
+}
+
+Type* HipPromoteIntsPass::getPromotedType(Type* TypeToPromote) {
+    if (auto* IntTy = dyn_cast<IntegerType>(TypeToPromote)) {
+        unsigned PromotedWidth = getPromotedBitWidth(IntTy->getBitWidth());
+        return Type::getIntNTy(TypeToPromote->getContext(), PromotedWidth);
+    }
+    return TypeToPromote;  // Return original type if not an integer
+}
+
+struct Replacement {
+    Instruction* Old;
+    Value* New;
+    Replacement(Instruction* O, Value* N) : Old(O), New(N) {}
+};
+
+void processInstruction(Instruction *I, Type *NonStdType, Type *PromotedTy, std::string Indent,
```

Comment:
Avoid a string copy?
```suggestion
void processInstruction(Instruction *I, Type *NonStdType, Type *PromotedTy, const std::string &Indent,
```